docs/userguide/src/migration/template.md

```diff
@@ -0,0 +1,127 @@
+# The template of the API migration guide
+
+This template is for the developers of MMTk-core.
+
+
+## View control
+
+The following buttons allow the readers to choose how many details they want to read.  Therefore, we
+can keep the TL;DR and the first two levels of the lists terse, and add detailed explanations at the
+third level.
+
+Try clicking those buttons and see what they do.
+
+{{#include ../../assets/snippets/view-controls.html}}
+
+<div id="api-migration-detail-body"><!-- We use JavaScript to process things within this div. -->
+
+## 0.xx.0
+
+### Title of a change
+
+```admonish tldr
+Use a few sentences to summarize the change so that the reader knows what has been changed without
+reading through the following list.
+
+Keep in mind that the reader can use the buttons above to hide everything but the TL;DR.  Make sure
+the TL;DR part covers all types/modules that are changed so that the readers know what changed by
+reading the TL;DR alone.
+```
+
+API changes:
+
+*   type `Foo`
+    -   `abc()` is removed.
+        +   Use `abc2()` instead.
+        +   *(Note 1: Put types/modules/... on the first level.  Keep the prefix "type", "module",
+            ...  before `Foo` so that readers can search for "type Foo" to find actual changes on
+            `Foo` instead of places that merely mention `Foo`.)*
+        +   *(Note 2: Put functions/methods/constants/... on the second level.)*
+        +   *(Note 3: Put suggestions and more details on the third level.)*
+    -   `defg()`
+        +   It now takes a new argument `foobarbaz` which requires blah blah...
+        +   And you need to blah blah blah blah blah blah blah blah blah blah blah blah blah blah
+            blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah
+            blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah...
+        +   *(Note 1: If the change is more complicated than "... is removed", feel free to put it
+            down one level.)*
+        +   *(Note 2: Since the third level is collapsible, feel free to add more details.)*
+*   module `aaa::bbb::ccc`
+    -   **Only affects users of feature "xxxx"** *(Note: When omitted, it affects everyone.)*
+    -   `method1()`
+        +   What happened to it...
+        +   Suggestions...
+    -   `method2()`
+        +   What happened to it...
+        +   Suggestions...
+
+VM bindings need to re-implement the following traits:
```

Comment:
What about "Things changed in mmtk-core"?

I think "VM->MMTk" and "MMTk->VM" is more accurate, but the reader needs to think what the arrows mean, and it is not immediately clear what the reader should do about them.  "Things changed in mmtk-core" tells them that those things changed on the mmtk-core side, but we don't know where you called them, so make changes if you need; while "VM bindings need to reimplement ..." clearly tells them they must rewrite those functions to conform to the new signatures.

Reply:
> What about "Things changed in mmtk-core"?
> 
> I think "VM->MMTk" and "MMTk->VM" is more accurate, but the reader needs to think what the arrows mean, and it is not immediately clear what the reader should do about them. 

I cannot find proper terms that are both ambiguous and straightforward to describe these two kinds of APIs. We may ask suggestions from others. One possibility is to put the 'VM binding' part under the 'API change' part as `trait VMBinding` -- maybe it is more proper.

> "Things changed in mmtk-core" tells them that those things changed on the mmtk-core side, but we don't know where you called them, so make changes if you need; while "VM bindings need to reimplement ..." clearly tells them they must rewrite those functions to conform to the new signatures.

In this case, you are talking about whether an API provided by MMTk is mandatory or optional for a binding. "VM->MMTK" API can be mandatory (e.g. `gc_init`), or optional (e.g. `gc_poll`). "MMTK->VM" API can be mandatory (e.g. `Collection::stop_all_mutators`), or optional (e.g. `Collection::post_forwarding`). Also the 'only affects users' note should cover this.

Reply:
>  In this case, you are talking about whether an API provided by MMTk is mandatory or optional for a binding.

No.  I am not.  It is a "VM calls this" (MMTk implements it) vs "VS implements this" (MMTk calls it) dichotomy.  The accurate term is still `VM -> MMTk` and `MMTk -> VM`, but I am describing them from the perspective of the VM.

I am not talking about whether an API is mandatory or optional.  For example, the VM must call `gc_init` somewhere, but MMTk core does not know where the VM calls it because the concrete place to initialize MMTk is VM-specific.  So we can't tell the VM binding to "change `MMTkHeap::initialize()` and call `gc_init` differently" because that's OpenJDK-specific.  I can only tell the VM binding "`gc_init` changed, and you call it differently wherever you call it".  On the contrary, `Scanning` is a trait defined by mmtk-core, so we can just tell the VM binding "Reimplement `Scanning`!"

Reply:
We could use 'downcall API' vs 'upcall API'. We may ask Steve for some suggestions.

Reply:
I removed this sub-title.